### PR TITLE
ItemPointer of AO tables need to be fixed

### PR DESCRIPTION
**Pull request by zhangjackey**
We generate the fake ItemPointer for AO table's tuple, but it is
a little different from the HEAP table's ItemPointer, the Offset
of the pointer can be 0 when it is a fake ItemPointer, but we
treat 0 to be invalid in the previous codes, In order to run
correctly, we set the 16th bit of Offset to be 1 as a flag, then
it can pass some check, but the value is not the real value, so when
using the fake ItemPointer, we have to note to process the flag.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by zhangjackey** [comment]
It can solve the Issue https://github.com/greenplum-db/gpdb/issues/7039 .

**Comment by hlinnaka** [comment]
I don't understand what the problem is. Does bitmap index AM get confused by offset number 32768? How, why?

**Comment by pivotal-ning-yu** [comment]
Copy Heikki's reply from #7039:

> Ah, ok, looking at the github issue explained that. So the problem isn't with 32768 as such, the problem is that the bitmap index build code expects to see the tuple IDs in order. It gets confused when it sees offset number 32768 before offset number 1.
>
> I guess this patch is an ok way to fix that, but I'd like to see some more sanity checks in that macro. Can we add an assertion or elog there, to check that the offset number is always in the range 1..32768. Because if we ever had offset number >= 32769, then things would go wrong. Likewise, a genuine 0 would be bad, since we're now mapping 32768 to 0.

**Comment by zhangjackey** [comment]
@hlinnaka 
Thanks a lot for your comments,  I add an assertion before we call BM_IPTR_TO_INT.

File: src/include/access/bitmap.h [hunk 1]
```diff
@@ -344,11 +344,15 @@ typedef struct BMTidBuildBuf
 
 /*
  * Convert an ItemPointer to and from an integer representation
+ *
+ * NOTE: If the ItemPointer is generated by AO/AOCS tables, the
+ * value of ip_posid field may be 0x8000, we must process this
+ * case. For more details please see the note of AOTupleId struct.
```
Comment (hlinnaka) [inline src/include/access/bitmap.h]: Please explain in this comment, why we can't just let 0x8000 be 0x8000, and need to map it to 0 instead. (The reason is that the bitmap AM building code expects to see the TIDs in ascending order, while the AO code will return 0x8000 before 1-0xFFFF.)

**Pull request merged**